File: 846-hand-of-straights/846-hand-of-straights.swift
class Solution {
    func isNStraightHand(_ hand: [Int], _ groupSize: Int) -> Bool {
        if hand.count % groupSize != 0 {
            return false
        }
        
        var groupDict: [Int: Int] = [:]
        var sortHand: [Int] = hand.sorted()
        
        while !sortHand.isEmpty {
            let now: Int = sortHand[0]
            groupDict[now] = groupSize - 1
            
            for i in 1..<groupSize {
                let next: Int = now + i
                
                if !sortHand.contains(next) {
                    return false
                }
                
                groupDict[now]! -= 1
                guard let idx = sortHand.firstIndex(of: next) else { return false }
                sortHand.remove(at: idx)
            }
            
            sortHand.remove(at: 0)
        }
        
        return true
    }
}

File: 846-hand-of-straights/846-hand-of-straights.py
from collections import defaultdict

class Solution(object):

    def isNStraightHand(self, hand, groupSize):
        if len(hand) % groupSize != 0:
            return False 
        
        groupDict = defaultdict(int)
        hand.sort()

        while hand:
            now = hand[0]
            groupDict[now] = groupSize - 1

            for i in range(1, groupSize):
                next = now + i 

                if next not in hand:
                    return False 
                
                groupDict[now] -= 1
                hand.remove(next)
            
            hand.remove(now)
        
        return True 

                